File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/ConflictResolutionTest.java
package ru.tinkoff.kora.kora.app.annotation.processor;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class ConflictResolutionTest extends AbstractKoraAppTest {
    @Test
    public void testMultipleComponentCandidates() {
        var result = compile(List.of(new KoraAppProcessor()), """
            public interface TestInterface {}
            """, """
            public class TestImpl1 implements TestInterface {}
            """, """
            public class TestImpl2 implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default String root(TestInterface t) {return "";}
                default TestImpl1 testImpl1() { return new TestImpl1(); }
                default TestImpl2 testImpl2() { return new TestImpl2(); }
            }
            """);
        assertThat(result.isFailed()).isTrue();
    }

    @Test
    public void testDefaultComponentOverride() throws ClassNotFoundException {
        var draw = compile("""
            public interface TestInterface {}
            """, """
            public class TestImpl1 implements TestInterface {}
            """, """
            public class TestImpl2 implements TestInterface {}
            """, """
            @KoraApp
            public interface ExampleApplication {
                @Root
                default String root(TestInterface t) {return "";}
                @DefaultComponent
                default TestImpl1 testImpl1() { return new TestImpl1(); }
                default TestImpl2 testImpl2() { return new TestImpl2(); }
            }
            """);
        assertThat(draw.getNodes()).hasSize(2);
        var g = draw.init().block();
        assertThat(g.get(draw.getNodes().get(0))).isInstanceOf(this.compileResult.loadClass("TestImpl2"));
    }

}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/ConflictResolutionTest.kt
package ru.tinkoff.kora.kora.app.ksp

import com.google.devtools.ksp.processing.SymbolProcessorProvider
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.ksp.common.AbstractSymbolProcessorTest

class ConflictResolutionTest : AbstractSymbolProcessorTest() {
    @Test
    fun testMultipleComponentSameType() {
        compile(
            listOf<SymbolProcessorProvider>(KoraAppProcessorProvider()),
            """
            interface TestInterface            
            """.trimIndent(), """
            class TestImpl1 : TestInterface {}
            """.trimIndent(), """
            class TestImpl2 : TestInterface {}
            """.trimIndent(), """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface) = ""
                fun testImpl1() = TestImpl1()
                fun testImpl2() = TestImpl2()
            }
            
            """.trimIndent()
        )

        assertThat(compileResult.isFailed()).isTrue()
    }

    @Test
    fun testDefaultComponentOverride() {
        compile(
            listOf<SymbolProcessorProvider>(KoraAppProcessorProvider()),
            """
            interface TestInterface            
            """.trimIndent(), """
            class TestImpl1 : TestInterface {}
            """.trimIndent(), """
            class TestImpl2 : TestInterface {}
            """.trimIndent(), """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface) = ""
            
                fun testImpl1() = TestImpl1()

                @DefaultComponent
                fun testImpl2() = TestImpl2()
            }
            """.trimIndent()
        )

        assertThat(compileResult.isFailed()).isFalse()
    }

    @Test
    fun testDefaultComponentTemplateOverride() {
        compile(
            listOf<SymbolProcessorProvider>(KoraAppProcessorProvider()),
            """
            interface TestInterface <T>
            """.trimIndent(), """
            class TestImpl1 <T> : TestInterface <T> {}
            """.trimIndent(), """
            class TestImpl2 <T> : TestInterface <T> {}
            """.trimIndent(), """
            @KoraApp
            interface ExampleApplication {
                @Root
                fun root(t: TestInterface<String>) = ""
            
                fun <T> testImpl1() = TestImpl1<T>()

                @DefaultComponent
                fun <T> testImpl2() = TestImpl2<T>()
            }
            """.trimIndent()
        )

        compileResult.assertSuccess()
    }
}
